<div class="vorspann">Wer keine Ausnahme werfen kann und wem auch <i>final_action</i> (finally) von der Guideline Support Library nicht zur Verfügung steht, hat ein Problem. Ausnahmezustände verlangen Ausnahmeaktionen: <i>goto</i>. Wirklich?</div>
<div class="text">Um ehrlich zu sein, war ich sehr überrascht, dass die Guidelines <i>goto exit;</i> als letzte Rettung beschreiben. Hier sind die verbleibenden Regeln zur Fehlerbehandlung der C++ Core Guidelines.</div>
<div class="text"><ul><li> <a href="http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-no-throw-raii" class="" alt="%7B%22user_params%22%3A%22%22%2C%22version%22%3A1%2C%22destination%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re-no-throw-raii%22%2C%22type%22%3A%22E%22%2C%22subject%22%3A%22%22%2C%22custom%22%3A%7B%7D%2C%22text%22%3A%22E.25%3A%20If%20you%20can%C2%92t%20throw%20exceptions%2C%20simulate%20RAII%20for%20resource%20management%22%2C%22href%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re-no-throw-raii%22%2C%22anchor%22%3A%22%22%2C%22target%22%3A%22_blank%22%2C%22ir_link%22%3A1%2C%22alias%22%3A%22%22%7D" title="Link auf http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-no-throw-raii">E.25: If you can’t throw exceptions, simulate RAII for resource management</a> </li><li> <a href="http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-no-throw-crash" title="Link auf http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-no-throw-crash" class="" alt="%7B%22type%22%3A%22E%22%2C%22destination%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re-no-throw-crash%22%2C%22subject%22%3A%22%22%2C%22user_params%22%3A%22%22%2C%22version%22%3A1%2C%22text%22%3A%22E.26%3A%20If%20you%20can%C2%92t%20throw%20exceptions%2C%20consider%20failing%20fast%22%2C%22custom%22%3A%7B%7D%2C%22anchor%22%3A%22%22%2C%22href%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re-no-throw-crash%22%2C%22alias%22%3A%22%22%2C%22ir_link%22%3A1%2C%22target%22%3A%22_blank%22%7D">E.26: If you can’t throw exceptions, consider failing fast</a> </li><li> <a title="Link auf http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-no-throw-codes" class="" alt="%7B%22anchor%22%3A%22%22%2C%22href%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re-no-throw-codes%22%2C%22ir_link%22%3A1%2C%22alias%22%3A%22%22%2C%22target%22%3A%22_blank%22%2C%22subject%22%3A%22%22%2C%22type%22%3A%22E%22%2C%22destination%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re-no-throw-codes%22%2C%22version%22%3A1%2C%22user_params%22%3A%22%22%2C%22custom%22%3A%7B%7D%2C%22text%22%3A%22E.27%3A%20If%20you%20can%C2%92t%20throw%20exceptions%2C%20use%20error%20codes%20systematically%22%7D" href="http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-no-throw-codes">E.27: If you can’t throw exceptions, use error codes systematically</a> </li><li> <a title="Link auf http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-specifications" alt="%7B%22anchor%22%3A%22%22%2C%22href%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re-specifications%22%2C%22alias%22%3A%22%22%2C%22ir_link%22%3A1%2C%22target%22%3A%22_blank%22%2C%22subject%22%3A%22%22%2C%22type%22%3A%22E%22%2C%22destination%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re-specifications%22%2C%22version%22%3A1%2C%22user_params%22%3A%22%22%2C%22text%22%3A%22E.30%3A%20Don%C2%92t%20use%20exception%20specifications%22%2C%22custom%22%3A%7B%7D%7D" class="" href="http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-specifications">E.30: Don’t use exception specifications</a> </li><li> <a class="" alt="%7B%22subject%22%3A%22%22%2C%22type%22%3A%22E%22%2C%22destination%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re_catch%22%2C%22version%22%3A1%2C%22user_params%22%3A%22%22%2C%22custom%22%3A%7B%7D%2C%22text%22%3A%22E.31%3A%20Properly%20order%20your%20catch-clauses%22%2C%22anchor%22%3A%22%22%2C%22href%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re_catch%22%2C%22alias%22%3A%22%22%2C%22ir_link%22%3A1%2C%22target%22%3A%22_blank%22%7D" title="Link auf http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re_catch" href="http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re_catch">E.31: Properly order your catch-clauses</a></li></ul></div>
<div class="text">Da die ersten drei Regeln stark zusammenhängen, werde ich sie zusammen beschreiben.</div>
<div class="ztitel"><ul><li><a title="Link auf http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-no-throw-raii" class="" alt="%7B%22href%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re-no-throw-raii%22%2C%22anchor%22%3A%22%22%2C%22target%22%3A%22_blank%22%2C%22ir_link%22%3A1%2C%22alias%22%3A%22%22%2C%22version%22%3A1%2C%22user_params%22%3A%22%22%2C%22subject%22%3A%22%22%2C%22type%22%3A%22E%22%2C%22destination%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re-no-throw-raii%22%2C%22custom%22%3A%7B%7D%2C%22text%22%3A%22E.25%3A%20If%20you%20can%27t%20throw%20exceptions%2C%20simulate%20RAII%20for%20resource%20management%22%7D" href="http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-no-throw-raii">E.25: If you can't throw exceptions, simulate RAII for resource management</a>, </li><li><a alt="%7B%22ir_link%22%3A1%2C%22alias%22%3A%22%22%2C%22target%22%3A%22_blank%22%2C%22anchor%22%3A%22%22%2C%22href%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re-no-throw-crash%22%2C%22custom%22%3A%7B%7D%2C%22text%22%3A%22E.26%3A%20If%20you%20can%C2%92t%20throw%20exceptions%2C%20consider%20failing%20fast%22%2C%22type%22%3A%22E%22%2C%22destination%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re-no-throw-crash%22%2C%22subject%22%3A%22%22%2C%22user_params%22%3A%22%22%2C%22version%22%3A1%7D" class="" title="Link auf http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-no-throw-crash" href="http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-no-throw-crash">E.26: If you can’t throw exceptions, consider failing fast</a> und </li><li><a title="Link auf http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-no-throw-codes" class="" alt="%7B%22version%22%3A1%2C%22user_params%22%3A%22%22%2C%22subject%22%3A%22%22%2C%22type%22%3A%22E%22%2C%22destination%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re-no-throw-codes%22%2C%22text%22%3A%22E.27%3A%20If%20you%20can%C2%92t%20throw%20exceptions%2C%20use%20error%20codes%20systematically%22%2C%22custom%22%3A%7B%7D%2C%22href%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re-no-throw-codes%22%2C%22anchor%22%3A%22%22%2C%22target%22%3A%22_blank%22%2C%22alias%22%3A%22%22%2C%22ir_link%22%3A1%7D" href="http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-no-throw-codes">E.27: If you can’t throw exceptions, use error codes systematically</a></li></ul></div>
<div class="text">Die zentrale Idee von RAII ist recht einfach. Wenn du auf eine Ressource aufpassen sollst, verpacke diese in eine Klasse. Verwende den Konstruktor der Klasse zum Initialisieren und den Destruktor der Klasse zum Freigeben der Ressource. Falls du nun ein lokales Objekt der Klasse auf dem Stack erzeugst, passt die C++-Laufzeit automatisch auf die Ressource und du bist die Verantwortung los. Mehr Details zu RAII gibt es in meinem Artikel "<a href="https://www.grimm-jaud.de/index.php/blog/garbage-collection-nein-danke" title="Link auf https://www.grimm-jaud.de/index.php/blog/garbage-collection-nein-danke" alt="%7B%22text%22%3A%22Garbage%20Collection%20%5B--%5D%20No%20Thanks%22%2C%22custom%22%3A%7B%7D%2C%22version%22%3A1%2C%22user_params%22%3A%22%22%2C%22subject%22%3A%22%22%2C%22type%22%3A%22E%22%2C%22destination%22%3A%22https%3A%2F%2Fwww.grimm-jaud.de%2Findex.php%2Fblog%2Fgarbage-collection-nein-danke%22%2C%22target%22%3A%22_blank%22%2C%22ir_link%22%3A1%2C%22alias%22%3A%22%22%2C%22href%22%3A%22https%3A%2F%2Fwww.grimm-jaud.de%2Findex.php%2Fblog%2Fgarbage-collection-nein-danke%22%2C%22anchor%22%3A%22%22%7D" class="">Garbage Collection [--] No Thanks</a>".</div>
<div class="text">Was heißt es nun, RAII für Ressourcenmanagement zu simulieren? Stelle dir vor, du hast eine Funktion [i]func[/i], die eine Ausnahme wirft, wenn <i>Gadget</i> nicht erzeugt werden kann: </div>
<div class="pre">void func(zstring arg)<br>{<br>&nbsp;&nbsp;&nbsp; Gadget g {arg};<br>&nbsp;&nbsp;&nbsp; // ...<br>}</div>
<div class="text">Wenn du keine Ausnahme werfen kannst, lässt sich RAII simulieren, indem du eine Methode <i>valid</i> für <i>Gadget</i> implementierst:</div>
<div class="pre">error_indicator func(zstring arg)<br>{<br>&nbsp;&nbsp;&nbsp; Gadget g {arg};<br>&nbsp;&nbsp;&nbsp; if (!g.valid()) return gadget_construction_error;<br>&nbsp;&nbsp;&nbsp; // ...<br>&nbsp;&nbsp;&nbsp; return 0;&nbsp;&nbsp; // zero indicates "good"<br>}</div>
<div class="text">In diesem Fall muss der Aufrufer der Funktion deren Rückgabewert prüfen.</div>
<div class="text">Die Regel E.26 ist sehr naheliegend. Falls es keine Möglichkeit wie beim Speichermangel gibt, dass sich das Programm von einem Fehler erholt, sollte es schnell fehlschlagen. Falls du keine Ausnahme in diesem Fall werfen kannst, rufe <a class="" alt="%7B%22custom%22%3A%7B%7D%2C%22text%22%3A%22std%3A%3Aabort%22%2C%22version%22%3A1%2C%22user_params%22%3A%22%22%2C%22subject%22%3A%22%22%2C%22destination%22%3A%22https%3A%2F%2Fen.cppreference.com%2Fw%2Fcpp%2Futility%2Fprogram%2Fabort%22%2C%22type%22%3A%22E%22%2C%22target%22%3A%22_blank%22%2C%22ir_link%22%3A1%2C%22alias%22%3A%22%22%2C%22href%22%3A%22https%3A%2F%2Fen.cppreference.com%2Fw%2Fcpp%2Futility%2Fprogram%2Fabort%22%2C%22anchor%22%3A%22%22%7D" title="Link auf https://en.cppreference.com/w/cpp/utility/program/abort" href="https://en.cppreference.com/w/cpp/utility/program/abort">std::abort</a> auf. Dies führt zur <i>abnormal</i>-Programmbeendigung:</div>
<div class="pre">void f(int n)<br>{<br>&nbsp;&nbsp;&nbsp; // ...<br>&nbsp;&nbsp;&nbsp; p = static_cast&lt;X*&gt;(malloc(n, X));<br>&nbsp;&nbsp;&nbsp; if (!p) abort();&nbsp;&nbsp;&nbsp;&nbsp; // abort if memory is exhausted<br>&nbsp;&nbsp;&nbsp; // ...<br>}</div>
<div class="text">Die <i>abnormal</i>-Programmbeendigung wird dann auftreten, wenn du keinen Signal Handler installierst hast, der das Signal <a alt="%7B%22custom%22%3A%7B%7D%2C%22text%22%3A%22SIBABRT%22%2C%22subject%22%3A%22%22%2C%22destination%22%3A%22https%3A%2F%2Fen.cppreference.com%2Fw%2Fcpp%2Futility%2Fprogram%2FSIG_types%22%2C%22type%22%3A%22E%22%2C%22version%22%3A1%2C%22user_params%22%3A%22%22%2C%22ir_link%22%3A1%2C%22alias%22%3A%22%22%2C%22target%22%3A%22_blank%22%2C%22anchor%22%3A%22%22%2C%22href%22%3A%22https%3A%2F%2Fen.cppreference.com%2Fw%2Fcpp%2Futility%2Fprogram%2FSIG_types%22%7D" class="" title="Link auf https://en.cppreference.com/w/cpp/utility/program/SIG_types" href="https://en.cppreference.com/w/cpp/utility/program/SIG_types">SIBABRT</a> fängt.</div>
<div class="text">Die Funktion [i]f[/i] verhält sich äquivalent zur folgenden Funktion:</div>
<div class="pre">void f(int n)<br>{<br>&nbsp;&nbsp;&nbsp; // ...<br>&nbsp;&nbsp;&nbsp; p = new X[n];&nbsp;&nbsp;&nbsp; // throw if memory is exhausted (by default, terminate)<br>&nbsp;&nbsp;&nbsp; // ...<br>}</div>
<div class="text">Jetzt schreibe ich über das Unwort [i]goto[/i] in der Regel E.27.</div>
<div class="text">Im Falle eines Fehlers müssen die folgenden Probleme gemäß der Guidelines gelöst werden:<br></div>
<div class="text"><ol><li>Wie signalisiert die Funktion ihren Fehlerfall?</li><li>Wie können alle Ressourcen freigegeben werden, bevor die Funktion beendet wird?</li><li>Was soll als Fehlerindikator verwendet werden?</li></ol></div>
<div class="text">Im Allgemeinen sollte deine Funktion zwei Rückgabewerte besitzen: den Wert und den Fehlerindikator. Dies lässt sich schön mit [i]std::pair[/i] abbilden. Das Freigeben der Ressourcen wird aber sehr schnell zum Wartungsalbtraum. Dies gilt selbst dann, wenn die Aufräumarbeit in Funktionen gekapselt wird:<br></div>
<div class="pre">std::pair&lt;int, error_indicator&gt; user()<br>{<br>&nbsp;&nbsp;&nbsp; Gadget g1 = make_gadget(17);<br>&nbsp;&nbsp;&nbsp; if (!g1.valid()) {<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; return {0, g1_error};<br>&nbsp;&nbsp;&nbsp; }<br><br>&nbsp;&nbsp;&nbsp; Gadget g2 = make_gadget(17);<br>&nbsp;&nbsp;&nbsp; if (!g2.valid()) {<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; cleanup(g1);<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; return {0, g2_error};<br>&nbsp;&nbsp;&nbsp; }<br><br>&nbsp;&nbsp;&nbsp; // ...<br><br>&nbsp;&nbsp;&nbsp; if (all_foobar(g1, g2)) {<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; cleanup(g1);<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; cleanup(g2);<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; return {0, foobar_error};<br>&nbsp;&nbsp;&nbsp; // ...<br><br>&nbsp;&nbsp;&nbsp; cleanup(g1);<br>&nbsp;&nbsp;&nbsp; cleanup(g2);<br>&nbsp;&nbsp;&nbsp; return {res, 0};<br>}<br></div>
<div class="text">Schaut doch gut aus. Oder?</div>
<div class="text">Weißt du, für was DRY steht? Don't Repeat Yourself. Obwohl in dem Beispiel die Aufräumarbeit in Funktionen verpackt ist, besitzt der Code ein deutliches Geschmäckle, da die Funktionen mehrfach an verschieden Stellen aufgerufen werden. Wie lässt sich die Wiederholung vermeiden? Verschiebe die Aufräumarbeit an das Ende der Funktion und springe hin:<br></div>
<div class="pre">std::pair&lt;int, error_indicator&gt; user()<br>{<br>&nbsp;&nbsp;&nbsp; error_indicator err = 0;<br><br>&nbsp;&nbsp;&nbsp; Gadget g1 = make_gadget(17);<br>&nbsp;&nbsp;&nbsp; if (!g1.valid()) {<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; err = g1_error;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; // (1)<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; goto exit;<br>&nbsp;&nbsp;&nbsp; }<br><br>&nbsp;&nbsp;&nbsp; Gadget g2 = make_gadget(17);<br>&nbsp;&nbsp;&nbsp; if (!g2.valid()) {<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; err = g2_error;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; // (1)<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; goto exit;<br>&nbsp;&nbsp;&nbsp; }<br><br>&nbsp;&nbsp;&nbsp; if (all_foobar(g1, g2)) {<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; err = foobar_error;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; // (1)<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; goto exit;<br>&nbsp;&nbsp;&nbsp; }<br>&nbsp;&nbsp;&nbsp; // ...<br><br>exit:<br>&nbsp; if (g1.valid()) cleanup(g1);<br>&nbsp; if (g2.valid()) cleanup(g2);<br>&nbsp; return {res, err};<br>}</div>
<div class="text">Zugegeben, mit der Hilfe von [i]goto[/i] wird die Struktur der Funktion deutlich übersichtlicher. Im Falle eines Fehlers wird lediglich der Fehlerindikator (1) gesetzt. Ausnahmezustände verlangen Ausnahmeaktionen. </div>
<div class="ztitel"><a href="http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-specifications" alt="%7B%22alias%22%3A%22%22%2C%22ir_link%22%3A1%2C%22target%22%3A%22_blank%22%2C%22anchor%22%3A%22%22%2C%22href%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re-specifications%22%2C%22text%22%3A%22E.30%3A%20Don%27t%20use%20exception%20specifications%22%2C%22custom%22%3A%7B%7D%2C%22subject%22%3A%22%22%2C%22destination%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re-specifications%22%2C%22type%22%3A%22E%22%2C%22version%22%3A1%2C%22user_params%22%3A%22%22%7D" class="" title="Link auf http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-specifications">E.30: Don't use exception specifications</a></div>
<div class="text">Zuerst einmal, hier ist ein Beispiel für eine Ausnahmespezifikation:<br></div>
<div class="pre">int use(int arg)<br>&nbsp;&nbsp;&nbsp; throw(X, Y)<br>{<br>&nbsp;&nbsp;&nbsp; // ...<br>&nbsp;&nbsp;&nbsp; auto x = f(arg);<br>&nbsp;&nbsp;&nbsp; // ...<br>}<br></div>
<div class="text">Das bedeutet, dass die Funktion use eine Ausnahme vom Typ X oder Y werfen kann. Falls eine andere Ausnahme geworfen wird, führt dies direkt zu [i]std::terminate[/i]. </div>
<div class="text">Die dynamische Ausnahmespezifikationen mit Argument [i]throw(X, Y)[/i] und ohne Argument throw() ist deprecated seit C++11. Die dynamische Ausnahmespezifikation mit Argument wird mit C++17, die ohne Argument mit C++20 entfernt. [i]throw()[/i] ist äquivalent zu [i]noexcept[/i]. Hier gibt es mehr Details dazu: "<a href="/exec/mainmenu.pl?sid=4fb30f3c5366c6a0e67c4b1e6899e4a5&amp;rm=open_article_id&amp;bid=2472518" title="Link auf Beitrag 2472518" alt="%7B%22anchor%22%3A%22%22%2C%22href%22%3A%22%2Fexec%2Fmainmenu.pl%3Fsid%3D4fb30f3c5366c6a0e67c4b1e6899e4a5%26rm%3Dopen_article_id%26bid%3D2472518%22%2C%22ir_link%22%3A1%2C%22alias%22%3A%22%22%2C%22target%22%3A%22_blank%22%2C%22destination%22%3A2472518%2C%22type%22%3A%22B%22%2C%22subject%22%3A%22%22%2C%22user_params%22%3A%22%22%2C%22version%22%3A1%2C%22custom%22%3A%7B%7D%2C%22text%22%3A%22C%2B%2B%20Core%20Guidelines%3A%20Der%20noexcept-Spezifier%20und%20-Operator%22%7D" class="">C++ Core Guidelines: Der noexcept-Spezifier und -Operator</a>".&nbsp; </div>
<div class="text">Wenn du die letzte Regel nicht kennst, kann dies zu unangenehmen Überraschungen führen. </div>
<div class="ztitel"><a title="Link auf http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re_catch" alt="%7B%22anchor%22%3A%22%22%2C%22href%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re_catch%22%2C%22ir_link%22%3A1%2C%22alias%22%3A%22%22%2C%22target%22%3A%22_blank%22%2C%22type%22%3A%22E%22%2C%22destination%22%3A%22http%3A%2F%2Fisocpp.github.io%2FCppCoreGuidelines%2FCppCoreGuidelines%23Re_catch%22%2C%22subject%22%3A%22%22%2C%22user_params%22%3A%22%22%2C%22version%22%3A1%2C%22text%22%3A%22E.31%3A%20Properly%20order%20your%20catch-clauses%22%2C%22custom%22%3A%7B%7D%7D" class="" href="http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re_catch">E.31: Properly order your catch-clauses</a></div>
<div class="text">Eine Ausnahme wird auf Grundlage der Best-fit-Strategie gefangen. Das heißt, die erste Ausnahmebehandlung, die zur aktuellen Ausnahme passt, wird verwendet. Dies ist der Grund, dass die Ausnahmebehandlung vom Speziellem zum Allgemeinem strukturiert werden soll. Wenn nicht, kommt die spezifische Ausnahmebehandlung nie zum Einsatz. Im folgenden Beispiel ist [i]DivisionByZeroException[/i] von [i]std::exception[/i] abgeleitet. <br></div>
<div class="pre">try{<br>&nbsp;&nbsp;&nbsp; // throw an exception&nbsp;&nbsp; (1) <br>}<br>catch(const DivisionByZeroException&amp; ex){ .... } // (2) <br>catch(const std::exception&amp; ex{ .... }&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; // (3) <br>catch(...){ .... }&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; // (4) <br></div>
<div class="text">In diesem Fall, kommt die Ausnahmebehandlung in [i]DivisionByZeroException[/i] (2) zuerst zum Einsatz, falls die Ausnahme (1) geworfen wird. Falls die spezielle Ausnahmebehandlung nicht passt, werde alle Ausnahme, die von [i]std::exception[/i] abgeleitet sind, in der folgenden Zeile (3) gefangen. Die letzte Ausnahmebehandlung besitzt eine Ellipse (4) und kann damit alle Ausnahme fangen. </div>
<div class="ztitel">Wie geht's weiter?</div>
<div class="text">Wie bereits angekündigt, werde ich mich im nächsten Artikel mit den fünf Regeln zur Konstanz und Immutablilität beschäftigen.</div>
<div class="ztitel">Die pdf-Päckchen sind fertig</div>
<div class="text"><ul><li>[De]: <a href="https://goo.gl/uxJZ4U" title="Link auf https://goo.gl/uxJZ4U" class="" alt="%7B%22custom%22%3A%7B%7D%2C%22text%22%3A%22Embedded%3A%20Performanz%20z%C3%A4hlt%22%2C%22version%22%3A1%2C%22user_params%22%3A%22%22%2C%22subject%22%3A%22%22%2C%22destination%22%3A%22https%3A%2F%2Fgoo.gl%2FuxJZ4U%22%2C%22type%22%3A%22E%22%2C%22target%22%3A%22_blank%22%2C%22alias%22%3A%22%22%2C%22ir_link%22%3A1%2C%22href%22%3A%22https%3A%2F%2Fgoo.gl%2FuxJZ4U%22%2C%22anchor%22%3A%22%22%7D">Embedded: Performanz zählt</a> <br></li><li>[En]: <a title="Link auf https://goo.gl/TGgjJ5" alt="%7B%22text%22%3A%22C%2B%2B%20Core%20Guidelines%3A%20Concurrency%20and%20Parallelism%22%2C%22custom%22%3A%7B%7D%2C%22type%22%3A%22E%22%2C%22destination%22%3A%22https%3A%2F%2Fgoo.gl%2FTGgjJ5%22%2C%22subject%22%3A%22%22%2C%22user_params%22%3A%22%22%2C%22version%22%3A1%2C%22alias%22%3A%22%22%2C%22ir_link%22%3A1%2C%22target%22%3A%22_blank%22%2C%22anchor%22%3A%22%22%2C%22href%22%3A%22https%3A%2F%2Fgoo.gl%2FTGgjJ5%22%7D" class="" href="https://goo.gl/TGgjJ5">C++ Core Guidelines: Concurrency and Parallelism</a><br></li></ul></div>
